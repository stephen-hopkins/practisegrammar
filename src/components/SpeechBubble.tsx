import { classNames } from "primereact/utils";
import styles from "./SpeechBubble.module.css";

type Props = {
  text: string;
  type: "original" | "translated";
  show: boolean;
};

export default function SpeechBubble({ text, type, show }: Props) {
  return (
    <div className={classNames({ hidden: !show }, type == "original" ? styles.right : styles.left, styles.bubble)}>
      <div className={styles.text_wrapper}>
        <span className={styles.text}>{text}</span>
      </div>
    </div>
  );
}
