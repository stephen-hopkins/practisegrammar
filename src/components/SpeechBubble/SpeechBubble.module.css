.bubble {
  background-size: cover;
  aspect-ratio: 1096 / 499;
  margin-top: 1rem;
  width: fit-content;
}

.right {
  background-image: url("../../../public/speechbubble-right.png");
}

.left {
  background-image: url("../../../public/speechbubble-left.png");
}

.text_wrapper {
  margin-top: 3%;
  height: 61%;
  display: flex;
  flex-flow: column;
  justify-content: center;
}

.text {
  margin-left: 2rem;
  margin-right: 2rem;
  color: black;
}
