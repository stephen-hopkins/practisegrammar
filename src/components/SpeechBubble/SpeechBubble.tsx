import { classNames } from "primereact/utils";
import styles from "./SpeechBubble.module.css";

type Props = {
  text: string;
  type: "original" | "translated";
  show: boolean;
  className: string;
  onClick: () => void;
};

export default function SpeechBubble({ text, type, show, className, onClick }: Props) {
  if (!show) {
    return null;
  }

  return (
    <div
      className={classNames(type == "original" ? styles.right : styles.left, styles.bubble, className)}
      onClick={onClick}
    >
      <div className={styles.text_wrapper}>
        <span className={styles.text}>{text}</span>
      </div>
    </div>
  );
}
