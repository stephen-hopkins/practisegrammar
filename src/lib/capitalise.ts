export function capitalise(word: string) {
  return word.charAt(0).toLocaleLowerCase() + word.slice(1);
}