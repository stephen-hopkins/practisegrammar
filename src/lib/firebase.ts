import { initializeApp } from "firebase/app";
import { Concept, Language } from "@/src/models/constants";
import { addDoc, collection, CollectionReference, getDocs, getFirestore, query, where } from "@firebase/firestore";
import { Sentence, SentenceInfo } from "@/src/models/translation";

const firebaseConfig = {
  apiKey: "AIzaSyCXVeg1hzBtWVW2FK9_cz7r7gygHIH8E2k",
  authDomain: "practisegrammar.firebaseapp.com",
  projectId: "practisegrammar",
  storageBucket: "practisegrammar.appspot.com",
  messagingSenderId: "666711626320",
  appId: "1:666711626320:web:410a88534639301d4b7755",
  measurementId: "G-YCLPJ0Z093",
};

function getDb() {
  // const analytics = getAnalytics(app);
  const app = initializeApp(firebaseConfig);
  const db = getFirestore(app);

  return {
    sentences: collection(db, "sentences") as CollectionReference<SentenceInfo>,
  };
}

export async function getSentence(language: Language, concept: Concept, word: string) {
  const { sentences } = getDb();
  const q = query(
    sentences,
    where("language", "==", language),
    where("concept", "==", concept),
    where("word", "==", word)
  );
  const querySnapshot = await getDocs(q);
  if (!querySnapshot.empty) {
    return querySnapshot.docs[0].data();
  }
}

export async function addSentence(sentence: Sentence) {
  const { translations, ...metaData } = sentence;
  const { sentences } = getDb();
  //const doc = await addDoc(sentences, metaData, "translations", translations);
  //await addDoc(collection(sentences, doc.id, "translations"), translations);
}
