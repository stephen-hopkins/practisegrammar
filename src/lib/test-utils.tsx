import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { RenderOptions, render } from "@testing-library/react";
import "@testing-library/jest-dom";

jest.mock("next/router", () => require("next-router-mock"));

const queryClient = new QueryClient();
const AllProviders = ({ children }: { children: React.ReactNode }) => {
  return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>;
};

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
