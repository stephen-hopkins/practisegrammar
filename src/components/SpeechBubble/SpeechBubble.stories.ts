import { Meta, StoryObj } from "@storybook/react";
import SpeechBubble from "./SpeechBubble";

const meta: Meta<typeof SpeechBubble> = {
  title: "SpeechBubble",
  component: SpeechBubble,
  argTypes: { onClick: { action: "clicked" } },
};

export default meta;

type Story = StoryObj<typeof SpeechBubble>;

export const Original: Story = {
  args: {
    text: "Hello World",
    type: "original",
    show: true,
  },
};

export const Translated: Story = {
  args: {
    ...Original.args,
    type: "translated",
  },
};

export const LongText: Story = {
  args: {
    ...Original.args,
    text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum iaculis tellus ut ligula tincidunt pellentesque. Integer sed mi et diam consectetur rutrum. Aenean ut scelerisque est.",
  },
};

export const ShowFalse: Story = {
  args: {
    ...Original.args,
    show: false,
  },
};
