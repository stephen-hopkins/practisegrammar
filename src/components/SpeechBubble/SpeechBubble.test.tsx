import { render, screen } from "@/src/lib/test-utils";
import SpeechBubble from "./SpeechBubble";
import userEvent from "@testing-library/user-event";

describe("SpeechBubble", () => {
  it("renders", () => {
    render(<SpeechBubble className="test" text="Example test" show={true} type="original" onClick={() => {}} />);
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <SpeechBubble className="test" text="Example test" show={true} type="original" onClick={() => {}} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with className", () => {
    const { container } = render(
      <SpeechBubble className="test" text="Example test" show={true} type="original" onClick={() => {}} />
    );
    expect(container.firstChild).toHaveClass("test");
  });

  it("shows text", () => {
    const { getByText } = render(
      <SpeechBubble className="test" text="Example test" show={true} type="original" onClick={() => {}} />
    );
    expect(getByText("Example test")).toBeInTheDocument();
  });

  it("does not show when show is false", () => {
    const { container } = render(
      <SpeechBubble className="test" text="Example test" show={false} type="original" onClick={() => {}} />
    );
    const text = screen.queryByText("Example test");
    expect(text).toBeNull();
  });

  it("fires event when clicked", async () => {
    const mockOnClick = jest.fn();
    const user = userEvent.setup();
    const { container } = render(
      <SpeechBubble className="test" text="Example test" show={true} type="original" onClick={mockOnClick} />
    );

    const text = screen.getByText("Example test");
    await user.click(text);

    expect(mockOnClick).toHaveBeenCalled();
  });
});
