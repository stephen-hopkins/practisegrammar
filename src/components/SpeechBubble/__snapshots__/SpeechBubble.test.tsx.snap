// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpeechBubble matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="right bubble test"
  >
    <div
      class="text_wrapper"
    >
      <span
        class="text"
      >
        Example test
      </span>
    </div>
  </div>
</DocumentFragment>
`;
